include/ginkgo/core/distributed/preconditioner/schwarz.hpp

```diff
@@ -0,0 +1,171 @@
+ * @ingroup precond
+ */
+namespace preconditioner {
+
+
+/**
+ * A Schwarz preconditioner is a simple domain decomposition preconditioner that
+ * generalizes the Block Jacobi preconditioner, incorporating options for
+ * different local subdomain solvers and overlaps between the subdomains.
+ *
+ * See Iterative Methods for Sparse Linear Systems (Y. Saad) for a general
+ * treatment and variations of the method.
+ *
+ * @note Currently overlaps are not supported (TODO).
+ *
+ * @tparam ValueType  precision of matrix elements
+ * @tparam IndexType  integral type of the preconditioner
+ *
+ * @ingroup schwarz
+ * @ingroup precond
+ * @ingroup LinOp
+ */
+template <typename ValueType = default_precision,
+          typename LocalIndexType = int32, typename GlobalIndexType = int64>
+class Schwarz
+    : public EnableLinOp<Schwarz<ValueType, LocalIndexType, GlobalIndexType>> {
+    friend class EnableLinOp<Schwarz>;
+    friend class EnablePolymorphicObject<Schwarz, LinOp>;
+
+public:
+    using EnableLinOp<Schwarz>::convert_to;
+    using EnableLinOp<Schwarz>::move_to;
+    using value_type = ValueType;
+    using index_type = GlobalIndexType;
+    using local_index_type = LocalIndexType;
+    using global_index_type = GlobalIndexType;
+
+    /**
+     * Returns the local system matrix (rank-local diagonal block).
+     *
+     * @return the local system matrix
+     */
+    std::shared_ptr<const LinOp> get_local_system_matrix() const
+    {
+        return local_system_matrix_;
+    }
+
+
+    GKO_CREATE_FACTORY_PARAMETERS(parameters, Factory)
+    {
+        /**
+         * Local solver factory.
+         */
+        std::shared_ptr<const LinOpFactory> GKO_FACTORY_PARAMETER_SCALAR(
+            local_solver_factory, nullptr);
+    };
+    GKO_ENABLE_LIN_OP_FACTORY(Schwarz, parameters, Factory);
+    GKO_ENABLE_BUILD_METHOD(Factory);
+
+protected:
+    /**
+     * Creates an empty Schwarz preconditioner.
+     *
+     * @param exec  the executor this object is assigned to
+     */
+    explicit Schwarz(std::shared_ptr<const Executor> exec)
+        : EnableLinOp<Schwarz>(std::move(exec))
+    {}
+
+    /**
+     * Creates a Schwarz preconditioner from a matrix using a Schwarz::Factory.
+     *
+     * @param factory  the factory to use to create the preconditoner
+     * @param system_matrix  the matrix this preconditioner should be created
+     *                       from
+     */
+    explicit Schwarz(const Factory* factory,
+                     std::shared_ptr<const LinOp> system_matrix)
+        : EnableLinOp<Schwarz>(factory->get_executor(),
+                               gko::transpose(system_matrix->get_size())),
+          parameters_{factory->get_parameters()},
+          local_system_matrix_{std::move(
+              as<experimental::distributed::Matrix<ValueType, LocalIndexType,
+                                                   GlobalIndexType>>(
+                  system_matrix.get())
+                  ->get_local_matrix())}
```

Comment:
I think it should keep the global one and pass the get_local_matrix when needed

Reply:
I think as this is just a local preconditioner, we should not need the global one atleast now. We can change that later in case we need it.

Reply:
~~I'm kind of torn here - by storing the local matrix only, the lifetime of the global system matrix and its components become disconnected, which makes reasoning about them harder. OTOH, once we start overlaps, the local matrix used for preconditioning becomes different, so we need to store it explicitly anyways.~~
Why are we storing the local system matrix at all? We could always get that from the local solver. To me, this seems to be mixing two levels of nesting in one.

Reply:
If you dont have a generated local solver, then you need to have the local matrix to generate the local solver from the passed in local solver factory. But your point regarding the disconnected components makes sense. I think for now, as @yhmtsai suggested, I will store the global matrix, and will have to do tha dynamic casts when I need the local matrix.

Reply:
Do you need it past the `generate` step? Because otherwise you only need to store the generated solver, just like a local preconditioner doesn't need to store the system matrix after generation ;)

Reply:
Yes, currently I dont. So, I will remove it altogether for now and we can add it later if needed.